spec/unit/puppet/provider/dsc_base_provider/dsc_base_provider_spec.rb

```diff
@@ -78,14 +79,129 @@
   end
 
   context '.canonicalize' do
-    it 'returns the specified resource if it is in the canonicalized resource cache'
-    it 'retrieves the resource from the machine if it is not in the cache'
-    it 'caches caches the specified resource if the retrieval returns nil'
-    it 'removes the dsc_psdscrunascredential from the resource if not specified'
-    it 'sets the dsc_psdscrunascredential to the value from the input, if specified'
-    it 'assigns the value of the discovered resource if it is a downcased match for the specified resource'
-    it 'assigns the value of the specified resource if it differs from the discovered resource by more than just casing'
-    it 'returns the array of canonicalized resources'
+    subject(:canonicalized_resource) { provider.canonicalize(context, [manifest_resource]) }
+
+    let(:resource_name_hash) { { name: 'foo', dsc_name: 'foo' } }
+    let(:namevar_keys) { %i[name dsc_name] }
+    let(:parameter_keys) { %i[dsc_parameter dsc_psdscrunascredential] }
+    let(:credential_hash) { { 'username' => 'foo', 'password' => 'bar' } }
+    let(:base_resource) { resource_name_hash.dup }
+
+    before(:each) do
+      allow(context).to receive(:debug)
+      allow(provider).to receive(:namevar_attributes).and_return(namevar_keys)
+      allow(provider).to receive(:fetch_cached_hashes).and_return(cached_canonicalized_resource)
+      described_class.class_variable_set(:@@cached_canonicalized_resource, []) # rubocop:disable Style/ClassVars
+    end
+
+    after(:all) do
```

Comment:
Shouldn't this be 

```suggestion
    after(:each) do
```

or are the test-cases so closely entangled that they need to spill over state?

Reply:
I can def swap this to be each, I think my original logic was we are resetting it in the `before(:each)` and only need to clear it again after they're all done.

Reply:
huh - I feel that's even more confusing; if there's something in at the start of the test it seems like that's already a spill-over from a different test? maybe elevate the `after(:each)` to the top-level and remove the duplicate initialisers?

Reply:
We could definitely move the before/after cleanup for the caching to the top scope before(:each) I think.

Reply:
Done!